app/upgrades/signetlaunch/upgrades.go

```diff
@@ -131,32 +117,96 @@ func upgradeFinalityParameters(
 	return k.SetParams(ctx, params)
 }
 
-// propLaunch runs the proposal of launch that is meant to insert new BTC Headers.
-func propLaunch(
+// upgradeLaunch runs the upgrade:
+// - Transfer ubbn funds for token distribution
+// - Insert new BTC Headers
+// - Insert new finality providers
+func upgradeLaunch(
 	ctx sdk.Context,
 	encCfg *appparams.EncodingConfig,
 	btcLigthK *btclightkeeper.Keeper,
 	btcStkK *btcstkkeeper.Keeper,
+	bankK bankkeeper.SendKeeper,
 ) error {
-	cdc := encCfg.Codec
+	if err := upgradeTokensDistribution(ctx, bankK); err != nil {
+		return err
+	}
 
-	newHeaders, err := LoadBTCHeadersFromData(cdc)
+	if err := upgradeBTCHeaders(ctx, encCfg.Codec, btcLigthK); err != nil {
+		return err
+	}
+
+	return upgradeSignedFPs(ctx, encCfg, btcStkK)
+}
+
+func upgradeTokensDistribution(ctx sdk.Context, bankK bankkeeper.SendKeeper) error {
+	data, err := LoadTokenDistributionFromData()
 	if err != nil {
 		return err
 	}
 
-	if err := insertBtcHeaders(ctx, btcLigthK, newHeaders); err != nil {
+	for _, td := range data.TokenDistribution {
+		receiver, err := sdk.AccAddressFromBech32(td.AddressReceiver)
+		if err != nil {
+			return err
+		}
+
+		sender, err := sdk.AccAddressFromBech32(td.AddressSender)
+		if err != nil {
+			return err
+		}
+
+		amount := sdk.NewCoin(appparams.BaseCoinUnit, sdkmath.NewInt(td.Amount))
+		if err := bankK.SendCoins(ctx, sender, receiver, sdk.NewCoins(amount)); err != nil {
+			return err
```

Comment:
Just few quesation for better understanding:
- what happens if this returns and error ? does upgrade stop ?
- I imagine `bankK.SendCoins`  does all necessary validation if sender have enough coins etc?

I am just thinking of the scenario if by mistake we have some invalid account as sender and what will happen then. I imagine upgrade will happen on chain which is priavate yet and it will fails, so we will be able easily to re-do it.

Reply:
yes, if it fails it is going to panic and stop the upgrade 
The upgrade is called during `PreBlocker` I expect that the context is handled with commit and rollback in case of failures, but in our case, since it is all private we will be able to run it again